editor/src/components/canvas/canvas-strategies/strategies/groups.spec.browser2.tsx

```diff
@@ -0,0 +1,1367 @@
+/* eslint jest/expect-expect: ["warn", { "assertFunctionNames": ["expect", "assertStylePropsSet"] }] */
+import { getSimpleAttributeAtPath } from '../../../../core/model/element-metadata-utils'
+import { BakedInStoryboardUID } from '../../../../core/model/scene-utils'
+import { getDomRectCenter } from '../../../../core/shared/dom-utils'
+import { defaultEither } from '../../../../core/shared/either'
+import { forceRight } from '../../../../core/shared/either'
+import { right } from '../../../../core/shared/either'
+import { foldEither } from '../../../../core/shared/either'
+import { mapEither } from '../../../../core/shared/either'
+import { fromString } from '../../../../core/shared/element-path'
+import { WindowPoint, windowPoint } from '../../../../core/shared/math-utils'
+import { forceNotNull } from '../../../../core/shared/optional-utils'
+import { create } from '../../../../core/shared/property-path'
+import { NO_OP } from '../../../../core/shared/utils'
+import { Modifiers, emptyModifiers } from '../../../../utils/modifiers'
+import { selectComponentsForTest, wait } from '../../../../utils/utils.test-utils'
+import { EdgePositionBottomRight } from '../../canvas-types'
+import { CanvasControlsContainerID } from '../../controls/new-canvas-controls'
+import { mouseDragFromPointWithDelta } from '../../event-helpers.test-utils'
+import { EditorRenderResult, TestAppUID } from '../../ui-jsx.test-utils'
+import { TestSceneUID } from '../../ui-jsx.test-utils'
+import { makeTestProjectCodeWithSnippet, renderTestEditorWithCode } from '../../ui-jsx.test-utils'
+import { resizeElement } from './absolute-resize.test-utils'
+
+const GroupPath = `${BakedInStoryboardUID}/${TestSceneUID}/${TestAppUID}:root-div/group`
+
+function makeCodeSnippetForGroups(code: string) {
+  return `
+  <div data-uid='root-div' style={{width: 400, height: 400, position: 'relative'}}>
+    ${code}
+  </div>
+`
+}
+
+async function renderProjectWithGroup(code: string) {
+  const editor = await renderTestEditorWithCode(
+    makeTestProjectCodeWithSnippet(makeCodeSnippetForGroups(code)),
+    'await-first-dom-report',
+  )
+
+  return editor
+}
+
+async function dragByPixels(
+  editor: EditorRenderResult,
+  delta: WindowPoint,
+  testid: string,
+  modifiers: Modifiers = emptyModifiers,
+) {
+  const targetElement = editor.renderedDOM.getByTestId(testid)
+  const targetElementBounds = targetElement.getBoundingClientRect()
+  const targetElementCenter = windowPoint(getDomRectCenter(targetElementBounds))
+  const canvasControlsLayer = editor.renderedDOM.getByTestId(CanvasControlsContainerID)
+
+  await mouseDragFromPointWithDelta(canvasControlsLayer, targetElementCenter, delta, {
+    modifiers,
+    midDragCallback: async () => {
+      NO_OP()
+    },
+  })
+}
+
+function assertStylePropsSet(
+  editor: EditorRenderResult,
+  elementPathStr: string,
+  stylePropsObjToMatch: Record<string, any>,
+) {
```

Comment:
This is such a nice way to check the result without printing or string interpolating the expected code. The only thing I'm trying to wrap my head around is if the `stylePropsObjToMatch` should be exhaustive (so that we can also ensure no rogue props have been added along the way)

Reply:
that's why I put the undefined checks in. so you can assert a pin is definitely not defined.

I would not make this exhaustive, because then we introduce the noise of background colors, position: absolute etc